### ISSUE TITLE
Can multiple images/caption be input into the model?

### ISSUE THREAD

**Issue by WalterWangRevo** [IDEA-Research/GroundingDINO#102]
Hello, I find that the forward of the model is "outputs = model(image[None], captions=[caption])", which means the shape of image is [1,C,H,W] and captions is [str] .  But when I try to make multiple images stacked to [N,C,H,W] , it cannot work . Also, I try to make captions to [str1,str2] , and not surprisingly it cannot work. It seems that the model can only accept one image and one caption from the released code. If in my situation, I want to test M images, and each image is with N captions, I need to forward MxN times which is time consuming.

**Comment by rentainhe** [comment]
> Hello, I find that the forward of the model is "outputs = model(image[None], captions=[caption])", which means the shape of image is [1,C,H,W] and captions is [str] . But when I try to make multiple images stacked to [N,C,H,W] , it cannot work . Also, I try to make captions to [str1,str2] , and not surprisingly it cannot work. It seems that the model can only accept one image and one caption from the released code. If in my situation, I want to test M images, and each image is with N captions, I need to forward MxN times which is time consuming.

We will try to update a batch inference API for the users

**Comment by ashrielbrian** [comment]
Seconding this - our use case would be to perform offline batch inference and maximise GPU throughput

**Comment by andyoung009** [comment]
Hello, it is a interesting work. Duo to the complex model architecture, how can we quickly and intuitively obtain the overall model architecture DAG and the dimensional information of input and output features? I tried tensorboard but not succeed because multiple inputs are involved. Thank you!

**Comment by yuwenmichael** [comment]
> Seconding this - our use case would be to perform offline batch inference and maximise GPU throughput for M images with the same caption.
> 
> @rentainhe is there a timeline for this?
> 
> Edit:
> 
> I managed to get a batch script to work in a somewhat hacky way:
> 
> Stack the images into a batch:
> 
> ```python
> images = torch.stack([load_image(img)[1] for img in img_paths])
> boxes, logits, phrases = predict_batch(
>         model=model,
>         images=images,
>         caption=TEXT_PROMPT,
>         box_threshold=BOX_TRESHOLD,
>         text_threshold=TEXT_TRESHOLD
>     )
> ```
> 
> You'll need to update the `load_image` func to not use the `RandomResize`. Inside `datasets/transforms.py`, add this class:
> 
> ```python
> class Resize(object):
>     def __init__(self, size):
>         assert isinstance(size, (list, tuple))
>         self.size = size
> 
>     def __call__(self, img, target=None):
>         return resize(img, target, self.size)
> ```
> 
> Inside `load_image` in `inference.py`, I hardcoded the resize to ensure every image in the batch is of the same size. This is hacky and probably (definitely) results in poorer performance.
> 
> ```python
> transform = T.Compose(
>         [
>             # T.RandomResize([800], max_size=1333),
>             # Added T.Resize to fix the resized image during batch inference
>             T.Resize((800, 1200)),
>             T.ToTensor(),
>             T.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
>         ]
>     )
> ```
> 
> Adapting the existing `predict` function:
> 
> ```python
> def predict_batch(
>         model,
>         images: torch.Tensor,
>         caption: str,
>         box_threshold: float,
>         text_threshold: float,
>         device: str = "cuda"
> ) -> Tuple[torch.Tensor, torch.Tensor, List[str]]:
>     caption = preprocess_caption(caption=caption)
> 
>     model = model.to(device)
>     image = images.to(device)
> 
>     print(f"Image shape: {image.shape}") # Image shape: torch.Size([num_batch, 3, 800, 1200])
>     with torch.no_grad():
>         outputs = model(image, captions=[caption for _ in range(len(images))]) # <------- I use the same caption for all the images for my use-case
> 
>     print(f'{outputs["pred_logits"].shape}') # torch.Size([num_batch, 900, 256]) 
>     print(f'{outputs["pred_boxes"].shape}') # torch.Size([num_batch, 900, 4])
>     prediction_logits = outputs["pred_logits"].cpu().sigmoid()[0]  # prediction_logits.shape = (nq, 256)
>     prediction_boxes = outputs["pred_boxes"].cpu()[0]  # prediction_boxes.shape = (nq, 4)
> 
>     mask = prediction_logits.max(dim=1)[0] > box_threshold
>     logits = prediction_logits[mask]  # logits.shape = (n, 256)
>     boxes = prediction_boxes[mask]  # boxes.shape = (n, 4)
> 
>     tokenizer = model.tokenizer
>     tokenized = tokenizer(caption)
> 
>     phrases = [
>         get_phrases_from_posmap(logit > text_threshold, tokenized, tokenizer).replace('.', '')
>         for logit
>         in logits
>     ]
> 
>     return boxes, logits.max(dim=1)[0], phrases
> ```
> 
> This gave me a roughly 18% improvement in latency over single image inference of a batch of 16 images.

I successfully implement the method you provided, if anyone needs it, you can reference from this [link](https://github.com/yuwenmichael/Grounding-DINO-Batch-Inference).

**Comment by Kun-Ming** [comment]
> > Seconding this - our use case would be to perform offline batch inference and maximise GPU throughput for M images with the same caption.
> > @rentainhe is there a timeline for this?
> > Edit:
> > I managed to get a batch script to work in a somewhat hacky way:
> > Stack the images into a batch:
> > ```python
> > images = torch.stack([load_image(img)[1] for img in img_paths])
> > boxes, logits, phrases = predict_batch(
> >         model=model,
> >         images=images,
> >         caption=TEXT_PROMPT,
> >         box_threshold=BOX_TRESHOLD,
> >         text_threshold=TEXT_TRESHOLD
> >     )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You'll need to update the `load_image` func to not use the `RandomResize`. Inside `datasets/transforms.py`, add this class:
> > ```python
> > class Resize(object):
> >     def __init__(self, size):
> >         assert isinstance(size, (list, tuple))
> >         self.size = size
> > 
> >     def __call__(self, img, target=None):
> >         return resize(img, target, self.size)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Inside `load_image` in `inference.py`, I hardcoded the resize to ensure every image in the batch is of the same size. This is hacky and probably (definitely) results in poorer performance.
> > ```python
> > transform = T.Compose(
> >         [
> >             # T.RandomResize([800], max_size=1333),
> >             # Added T.Resize to fix the resized image during batch inference
> >             T.Resize((800, 1200)),
> >             T.ToTensor(),
> >             T.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
> >         ]
> >     )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Adapting the existing `predict` function:
> > ```python
> > def predict_batch(
> >         model,
> >         images: torch.Tensor,
> >         caption: str,
> >         box_threshold: float,
> >         text_threshold: float,
> >         device: str = "cuda"
> > ) -> Tuple[torch.Tensor, torch.Tensor, List[str]]:
> >     caption = preprocess_caption(caption=caption)
> > 
> >     model = model.to(device)
> >     image = images.to(device)
> > 
> >     print(f"Image shape: {image.shape}") # Image shape: torch.Size([num_batch, 3, 800, 1200])
> >     with torch.no_grad():
> >         outputs = model(image, captions=[caption for _ in range(len(images))]) # <------- I use the same caption for all the images for my use-case
> > 
> >     print(f'{outputs["pred_logits"].shape}') # torch.Size([num_batch, 900, 256]) 
> >     print(f'{outputs["pred_boxes"].shape}') # torch.Size([num_batch, 900, 4])
> >     prediction_logits = outputs["pred_logits"].cpu().sigmoid()[0]  # prediction_logits.shape = (nq, 256)
> >     prediction_boxes = outputs["pred_boxes"].cpu()[0]  # prediction_boxes.shape = (nq, 4)
> > 
> >     mask = prediction_logits.max(dim=1)[0] > box_threshold
> >     logits = prediction_logits[mask]  # logits.shape = (n, 256)
> >     boxes = prediction_boxes[mask]  # boxes.shape = (n, 4)
> > 
> >     tokenizer = model.tokenizer
> >     tokenized = tokenizer(caption)
> > 
> >     phrases = [
> >         get_phrases_from_posmap(logit > text_threshold, tokenized, tokenizer).replace('.', '')
> >         for logit
> >         in logits
> >     ]
> > 
> >     return boxes, logits.max(dim=1)[0], phrases
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This gave me a roughly 18% improvement in latency over single image inference of a batch of 16 images.
> 
> I successfully implement the method you provided, if anyone needs it, you can reference from this [link](https://github.com/yuwenmichael/Grounding-DINO-Batch-Inference).

Hi @yuwenmichael @ashrielbrian, 
I use this hacky way to implement batch inference of GroundingDino. But I was wondering why still get index at 0 in these two lines:

```python3
prediction_logits = outputs["pred_logits"].cpu().sigmoid()[0]  # prediction_logits.shape = (nq, 256)
prediction_boxes = outputs["pred_boxes"].cpu()[0]  # prediction_boxes.shape = (nq, 4)
```
In my case, this only give me the result of first image. In order to get all results, I need to yield all results from outputs of model in predict function like this:
```python3
for i in range(outputs["pred_boxes"].shape[0]):
        prediction_logits = outputs["pred_logits"].cpu().sigmoid()[0]  # prediction_logits.shape = (nq, 256)
        prediction_boxes = outputs["pred_boxes"].cpu()[0]  # prediction_boxes.shape = (nq, 4)
        
        mask = prediction_logits.max(dim=1)[0] > box_threshold
        logits = prediction_logits[mask]  # logits.shape = (n, 256)
        boxes = prediction_boxes[mask]  # boxes.shape = (n, 4)

        tokenizer = model.tokenizer
        tokenized = tokenizer(caption)
        
        if remove_combined:
            sep_idx = [i for i in range(len(tokenized['input_ids'])) if tokenized['input_ids'][i] in [101, 102, 1012]]
            
            phrases = []
            for logit in logits:
                max_idx = logit.argmax()
                insert_idx = bisect.bisect_left(sep_idx, max_idx)
                right_idx = sep_idx[insert_idx]
                left_idx = sep_idx[insert_idx - 1]
                phrases.append(get_phrases_from_posmap(logit > text_threshold, tokenized, tokenizer, left_idx, right_idx).replace('.', ''))
        else:
            phrases = [
                get_phrases_from_posmap(logit > text_threshold, tokenized, tokenizer).replace('.', '')
                for logit
                in logits
            ]

        yield boxes, logits.max(dim=1)[0], phrases
```
Could you please check if your result is aligned with the input?

**Comment by Andyyoung0507** [comment]
@ashrielbrian @Kun-Ming @rentainhe @WalterWangRevo @yuwenmichael ，thanks for all your contributions.
I also tried to inference multiple images by this code,  I have a question about this issue. How to distinguish which picture the generated boxes belong to? Different pictures have different numbers of Bboxes? How to solve this trouble caused in the process of mutiple image objects annotation?

**Comment by Kun-Ming** [comment]
Hi @Andyyoung0507, I think you could try my code above. I return the prediction result by a for...yield, in each iteration it would return a result of one img.

**Comment by Andyyoung0507** [comment]
> Hi @Andyyoung0507, I think you could try my code above. I return the prediction result by a for...yield, in each iteration it would return a result of one img.
Thank you! That is a good idea. Multiple images inference really brings about some difficulties that I can't deal with.

**Comment by FiorenzoParascandolo1** [comment]
> Seconding this - our use case would be to perform offline batch inference and maximise GPU throughput for M images with the same caption.
> 
> @rentainhe is there a timeline for this?
> 
> Edit:
> 
> I managed to get a batch script to work in a somewhat hacky way:
> 
> Stack the images into a batch:
> 
> ```python
> images = torch.stack([load_image(img)[1] for img in img_paths])
> boxes, logits, phrases = predict_batch(
>         model=model,
>         images=images,
>         caption=TEXT_PROMPT,
>         box_threshold=BOX_TRESHOLD,
>         text_threshold=TEXT_TRESHOLD
>     )
> ```
> 
> You'll need to update the `load_image` func to not use the `RandomResize`. Inside `datasets/transforms.py`, add this class:
> 
> ```python
> class Resize(object):
>     def __init__(self, size):
>         assert isinstance(size, (list, tuple))
>         self.size = size
> 
>     def __call__(self, img, target=None):
>         return resize(img, target, self.size)
> ```
> 
> Inside `load_image` in `inference.py`, I hardcoded the resize to ensure every image in the batch is of the same size. This is hacky and probably (definitely) results in poorer performance.
> 
> ```python
> transform = T.Compose(
>         [
>             # T.RandomResize([800], max_size=1333),
>             # Added T.Resize to fix the resized image during batch inference
>             T.Resize((800, 1200)),
>             T.ToTensor(),
>             T.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
>         ]
>     )
> ```
> 
> Adapting the existing `predict` function:
> 
> ```python
> def predict_batch(
>         model,
>         images: torch.Tensor,
>         caption: str,
>         box_threshold: float,
>         text_threshold: float,
>         device: str = "cuda"
> ) -> Tuple[torch.Tensor, torch.Tensor, List[str]]:
>     caption = preprocess_caption(caption=caption)
> 
>     model = model.to(device)
>     image = images.to(device)
> 
>     print(f"Image shape: {image.shape}") # Image shape: torch.Size([num_batch, 3, 800, 1200])
>     with torch.no_grad():
>         outputs = model(image, captions=[caption for _ in range(len(images))]) # <------- I use the same caption for all the images for my use-case
> 
>     print(f'{outputs["pred_logits"].shape}') # torch.Size([num_batch, 900, 256]) 
>     print(f'{outputs["pred_boxes"].shape}') # torch.Size([num_batch, 900, 4])
>     prediction_logits = outputs["pred_logits"].cpu().sigmoid()[0]  # prediction_logits.shape = (nq, 256)
>     prediction_boxes = outputs["pred_boxes"].cpu()[0]  # prediction_boxes.shape = (nq, 4)
> 
>     mask = prediction_logits.max(dim=1)[0] > box_threshold
>     logits = prediction_logits[mask]  # logits.shape = (n, 256)
>     boxes = prediction_boxes[mask]  # boxes.shape = (n, 4)
> 
>     tokenizer = model.tokenizer
>     tokenized = tokenizer(caption)
> 
>     phrases = [
>         get_phrases_from_posmap(logit > text_threshold, tokenized, tokenizer).replace('.', '')
>         for logit
>         in logits
>     ]
> 
>     return boxes, logits.max(dim=1)[0], phrases
> ```
> 
> This gave me a roughly 18% improvement in latency over single image inference of a batch of 16 images.

Hy to everyone, 
this allows you to do inference with batch size > 1, however I have a problem with reproducibility. If a certain couple (image, text) has different position in the batch then the output is different. It seems that the elements of the batch are not processed independently...

**Comment by SinanAkkoyun** [comment]
@rentainhe > We will try to update a batch inference API for the users
How does one run GroundingDINO batched?

**Comment by andreazuna89** [comment]
Hi all,
Are there any news about the batch implementation for Grounding Dino? I have added the predict_batch inside the inference.py as @KunmingS has implemented but I get similar behavior you have highlighted. I have a list of images and I use the predict class on each image of the list; the resulting logits and bounding boxes are slightly different respect to the results of the same images stacked and passed inside the predict_batch function for batch inference. The differences are very small but they are still present. Any improvements about the batch implementation? @rentainhe 

Thanks a lot
